### ISSUE TITLE
Question: Recommends of a Provides

### ISSUE THREAD

**Issue by NathanKell** [KSP-CKAN/NetKAN#8615]
Thanks to the merging in of RSS-CanaveralHD, that mod is now on CKAN (thanks folks!). But that then brings up a question I have about chaining things together. Because while I want to rec it in certain cases in RSS, I do not want to rec it for low end users. Right now RSS depends on a provider of RSSTextures, where four different mods each provide what RSS seeks. Example: https://github.com/KSP-CKAN/NetKAN/blob/master/NetKAN/RSSTextures16K.netkan

What would happen if I set the 8k and 16k netkans to rec RSS-CanaveralHD, but not the 2k or 4k ones? If a user installs, say, RP-1 (which then deps RO, which deps RSS, which deps RSSTextures) will RSS-CanaveralHD be added to the recs list (and auto-checked) if the 8k or 16k RSS Textures are selected, and not if the others are, or is that too late in the process for recs to take effect?

**Comment by DasSkelett** [comment]
Recommendations of dependencies (or other recommendations) are ignored, only recommendations of user-selected mods are offered to the user. This is regardless of whether it's through virtual modules or real ones.
So unless the user explicitly selects one of the two RSSTextures  on the mod list before clicking apply, they wouldn't see the recommendation.

**Comment by HebaruSan** [comment]
My mantra is, "The metadata should be accurate." If you think that someone installing `RSSTextures8192` or `RSSTextures16K` would likely want `RSS-CanaveralHD`, then I'd say to just add the recommendation. If there are cases where it isn't pulled in by the current CKAN client, we can treat that as a bug or enhancement and maybe work on improving it later.

**Comment by NathanKell** [comment]
Oooh, thanks! I thought that was true of suggests, not recs. I'll police our rec structure to be sure we're doing that right.

I guess that brings up a second question. Since I can't have this selected based on which texture set, could I do something like:
RP-1 depends on Anyof (four metapackages, each a combo of RSS + an RSS textures + either RSSCanaveral or not). That anyof should trigger upstream of when RSS's depends (for RSS-Textures) is checked?

Basically I want the user flow to be, select RP-1, then select a level of graphical fidelity (using depends/provides, or depends/anyof? Either better?), which then selects an appropriate set of mods. The complication being that I want the selection of texture to be upstream of RSS's own requirement of a texture set.

Does that make sense?

**Comment by HebaruSan** [comment]
If a user selects to install just RSS itself instead of RP-1, would you want the same flow to play out? I would think so, which would suggest these releationships should branch off of RealSolarSystem somehow.

**Comment by NathanKell** [comment]
Given what you've said however, I'm not sure that's possible to do, since unless I am misunderstanding you, you 100% cannot have conditional recs, and RSS obviously does not _depend_ on RSSCanaveral-HD at any time, even if I want to rec it if and only if you install a hirez RSS texture pack.

**Comment by HebaruSan** [comment]
Sorry—would you want that, or wouldn't you?

**Comment by NathanKell** [comment]
Ah, apologies, that was a heck of a run-on sentence.

I would like it to be the case that when a user selects RSS, and they are then presented with the option of one out of four texture packs to install (because RSS deps the RSSTextures metapackage, and each texture package provides it and conflicts it), that if the user selects either the 8k or 16k package (the two highest-resolution packages), the user is then recommended to install RSSCanaveral-HD.
However, if I correctly interpreted what you said about recs, that is impossible.

**Comment by HebaruSan** [comment]
Thanks for the explanation. There's some subtlety involved here that make it difficult to get a good intuitive grasp; first, the prompt to choose a texture module would be handled by throwing a `TooManyModsProvide` exception, which effectively cancels out of the installation and restarts it with an altered change set that includes the user's choice of textures. So that might make us think that having a new change set might allow for a new opportunity to review the recommendations of that change set. However, that's not the case; the GUI prompts the user to pick recommendations when the install first starts, based on the initial change set only, and downstream changes do not cause the recommendations screen to appear again.

It might be good to re-purpose this issue to track the overall idea of improving the user experience of installing RSS. We could analyze what the user sees and what decisions they need to make to look for opportunities to improve CKAN, both for RSS but also for other mods with complex relationships. For example, I think you asked in Discord about whether there was any way to add some help text when the user is choosing a `provides` mod; while this is not supported currently, there may be some low hanging fruit along the lines of allowing something like a `comment` property within a relationship object that could be shown on screen when resolving that relationship:

```yaml
depends:
  - name: RSSTextures
    comment: Choose 8k or 16k to get Canaveral-HD as well
```

**Comment by NathanKell** [comment]
That makes sense, and yes just having a comment would be awesome!

I think it's probably cleaner if I make a new issue and link to this one however.
